### DISCUSSION TITLE
`formatter.lineEnding` based on the operating system.

### DISCUSSION THREAD

**Discussion by assawalhy** [biomejs/biome#3290] [Q&A] [answered]
I'm working in a team to deliver a project and chose Biome as the promising and very fast solution for formatting and linting. Thx for such a great tool. Some of the teammates are using Windows and some are using Linux and as we know windows uses `crlf` as the line ending and Linux (POSIX systems) uses `lf`.

So, configuring Biome to use one of them only won't help us and will result to changes in all files if formatter in one OS and then formatted again all files in another OS.

I don't know if there is a solution for my use case or not. Any help is appreciated!

**Comment by arendjr** [comment]
Are you using Git? I think it can be configured to enforce one type of line-ending so that people who check in from “the other OS” will still commit with the line-ending expected for the repository.

**Reply by assawalhy** [reply]
I know that feature of git. But the problem here happens when running `biome format --write .`  it will change all the files in the repo and will be shown in the unstaged area and when added to the staged files (using `git add .`) these changes to the line ending will be reset back.

I've tried this scenario and here is what happened:

The OS is Windows, and I added the configuration `formatter.lineEnding: "crlf"` after not setting it. The formatting command will change all the files to have `crlf` but when these files are added to the staged area, they will be changed back to original state which I think was `lf`.

```
$ pnpm biome format --write .
Formatted 215 files in 179ms. Fixed 215 files.
```

---

I'm experiencing a strange behavior. If I changed `formatter.lineEnding` from `crlf` to `lf` then formatted all files, all the files will be changes (`git add .` will not stage anything), and the second time I format all files no files are changed. The same happens when I change it from `lf` to `crlf`, the first formatting changes all files, `git add .` add nothing to the staged area, and the second formatting do nothing.

**Reply by assawalhy** [reply]
https://github.com/biomejs/biome/assets/42011920/4939a91b-bfeb-4703-a9c9-afd7c5577d30

**Reply by arendjr** [reply, marked as answer]
I think the solution here is to configure your IDE to use the line ending that the repository expects. Even though you’re on Windows, you can still use a different line ending.

The confusing `git add .` behavior you witness is due to the files having no real changes after the line endings are normalized.

**Reply by arendjr** [reply]
Did this answer your question? Happy to close the discussion if it's resolved.

**Reply by petejonesabbvie** [reply]
To add another perspective, this is an ongoing pain-point for me on a large codebase with developers split between Windows and macOS.  

I have not found that VS Code's settings matter in this case, presumably because the IDE doesn't have the opportunity to open and resave files that have been changed by a separate CLI process (but that's just me speculating).  

We use VS Code across platforms, and have settled on `"lineEnding": "lf"`.  Regardless of VS Code settings, we still experience the problem described above – when folks on Windows run `biome format`, they see hundreds of changes due to the line endings being updated by the formatter.  Once they stage the files in git, the changes vanish because git is happy to replace the line endings, also as described above.

We have treated this as a training issue for new developers ("Don't panic!  Just stage the changes"), but I'd love to have an "auto" option similar to prettier's that removes this requirement.

**Reply by DavidSouther** [reply]
I'm having this in a CI environment, where the Windows github runner seems to be converting them to crlf, which then makes Biome cranky. (I could of course not run biome on windows, but then I'd need to rework my CI scripts, when I'd rather an easier config setting somewhere)

https://github.com/DavidSouther/ailly/actions/runs/14944559834/job/41986224997?pr=160

(Resolved with https://github.com/actions/checkout/issues/135#issuecomment-602171132 if anyone comes along later)

**Comment by arty-name** [comment]
Without Biome, git is by default doing the right thing with line endings: when checking out files, it converts them to the OS default. This way no one has to configure anything, it just works ©

Biome really should do the same instead of re-creating long solved problems.

**Comment by madkarlsson** [comment]
Just ran into this as well, really annoying when I'm switching OS's

https://github.com/biomejs/biome/discussions/3290#discussioncomment-9915501

That is absolutely not the solution. The solution for a CLI tool should never be IDE-dependent, especially since running this in the terminal on all files in a repo would then have to make me use the IDE to go through them? That's a bit backwards if you ask me. A "preserve" or "auto" that checks for consistency rather than enforcing it would make this truly agnostic.

As arty-name above says. Git solves and other version control tools make sure this works.

**Reply by siketyan** [reply]
We already introduced `auto` value for the option in Biome 2.3
